### PR TITLE
DM-39876: Add bandaid for pydantic v2

### PR DESCRIPTION
**Pull request by timj**
Pydantic v2 breaks everything. Rather than pinning to <2 instead use the facility provided by v2 to import v1 APIs. This will give us some flexibility on timeline for fixing things to work only with v2.

## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

File: python/lsst/daf/butler/core/logging.py [hunk 1]
```diff
@@ -175,10 +179,10 @@ class ButlerLogRecord(BaseModel):
     filename: str
     pathname: str
     lineno: int
-    funcName: str | None
+    funcName: str | None = None
     process: int
     processName: str
-    exc_info: str | None
+    exc_info: str | None = None
     MDC: dict[str, str]
```
Comment (andy-slac) [inline python/lsst/daf/butler/core/logging.py:186]: Does this work with v1? I thought that pydantic (v1) does not allow non-default attributes after default ones (but I may be thinking of dataclasses)?

**Review by andy-slac — APPROVED** [review]
Looks good, one minor question.

File: python/lsst/daf/butler/core/logging.py (hunk 1 above)
Comment (timj) [reply python/lsst/daf/butler/core/logging.py:186]: I made that change first with the auto updater and it seemed to be okay. I am running Jenkins over night on everything so I'll see whether everything works in the morning.

**Comment by timj** [comment]
I'm not sure what to do about the mypy failures. Mypy doesn't know about pydantic.v1 and with pydantic v1 it complains about thta type not existing because it never falls back to the except block. With pydantic v2 installed it gets even more confused.

**Comment by timj** [comment]
In theory changing mypy.ini to use `pydantic.v1` with pydantic v2 and then adding `type: ignore` on the fallback imports does work. Unfortunately it doesn't work at the moment because fastapi pins pydantic to v1 so when we install that pydantic reverts to v1 and then mypy gets confused about the `pydantic.v1` import (which we have to do first), since mypy always picks up the first import it sees.

**Pull request merged**